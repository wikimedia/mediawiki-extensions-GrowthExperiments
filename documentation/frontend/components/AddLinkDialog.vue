<template>
	<!-- eslint-disable vue/no-v-model-argument -->
	<onboarding-dialog
		v-model:open="wrappedOpen"
		v-model="modelValue"
		:total-steps="3"
		:initial-step="1"
		class="ext-growthExperiments-AddLinkDialog"
		:show-paginator="true"
	>
		<template #title>
			Introduction
		</template>
		<template #headerbtntext>
			Skip all
		</template>

		<template #step1>
			<div class="ext-growthExperiments-AddLinkDialog__image">
				<img
					src="../../../images/addlink/onboarding-image1-ltr.svg"
					width="360"
					height="216"
					alt="">
			</div>
			<h5 class="ext-growthExperiments-AddLinkDialog__title">
				Adding links will help people learn faster.
			</h5>
			<div class="ext-growthExperiments-AddLinkDialog__text">
				<p>
					You will decide whether words in one Wikipedia article should link
					to other Wikipedia articles.
				</p>
				<div
					class="ext-growthExperiments-AddLinkDialog__text__label">
					Example sentence
				</div>
				<div
					class="
						ext-growthExperiments-AddLinkDialog__text__example">
					The moon is the only
					<!-- eslint-disable-next-line max-len -->
					<mark class="ext-growthExperiments-AddLinkDialog__text__example__highlighted">
						natural satellite
					</mark> that
					<mark class="ext-growthExperiments-AddLinkDialog__text__example__highlighted">
						orbits
					</mark> around the
					<mark class="ext-growthExperiments-AddLinkDialog__text__example__highlighted">
						Earth
					</mark>.
				</div>
				<p>
					No special knowledge about the article is needed to do this task.
				</p>
			</div>
		</template>

		<template #step2>
			<div class="ext-growthExperiments-AddLinkDialog__image">
				<img
					src="../../../images/addlink/onboarding-image2-ltr.svg"
					width="360"
					height="216"
					alt="">
			</div>
			<h5 class="ext-growthExperiments-AddLinkDialog__title">
				Suggested links are machine-generated, and can be incorrect.
			</h5>
			<div class="ext-growthExperiments-AddLinkDialog__text">
				<p>
					<!-- eslint-disable-next-line max-len -->
					The suggestions might be on words that donâ€™t need them, or might link to the wrong article. Use your judgment to decide whether they are right or wrong.
				</p>

				<a class="ext-growthExperiments-AddLinkDialog__text__link" href="">
					Learn more about machine suggestions
				</a>
			</div>
		</template>
		<template #step3>
			<div class="ext-growthExperiments-AddLinkDialog__image">
				<img
					src="../../../images/addlink/onboarding-image3-ltr.svg"
					width="360"
					height="216"
					alt="">
			</div>
			<h5 class="ext-growthExperiments-AddLinkDialog__title">
				Guidelines
			</h5>
			<div class="ext-growthExperiments-AddLinkDialog__text">
				<div class="ext-growthExperiments-AddLinkDialog__text__list">
					<ul>
						<li>Link concepts that a reader might want to learn more about.</li>
						<li>Make sure the link is going to the right article.</li>
						<li>Don't link common words, years, or dates.</li>
						<li>If you're not sure, skip.</li>
					</ul>
				</div>
			</div>
		</template>

		<template #checkbox>
			Don't show again
		</template>
		<template
			#last-step-button-text>
			Get started
		</template>
	</onboarding-dialog>
</template>

<script>
import { ref, toRef } from 'vue';
import { useModelWrapper } from '@wikimedia/codex';
import OnboardingDialog from './OnboardingDialog.vue';

export default {
	name: 'AddLinkDialog',
	components: {
		OnboardingDialog
	},
	props: {
		/**
		 * Whether the dialog is visible. Should be provided via a v-model:open
		 * binding in the parent scope.
		 */
		open: {
			type: Boolean,
			default: false
		}

	},

	emits: [ 'update:open' ],

	setup( props, { emit } ) {
		const wrappedOpen = useModelWrapper( toRef( props, 'open' ), emit, 'update:open' );
		const modelValue = ref( false );

		return {
			modelValue,
			wrappedOpen
		};
	}
};
</script>

<style lang="less">
@import '../node_modules/@wikimedia/codex-design-tokens/dist/theme-wikimedia-ui.less';
@import '../node_modules/@wikimedia/codex/dist/mixins/link.less';
@import './variables.less';

.ext-growthExperiments-AddLinkDialog {
	width: @onboardingDialogWidth;
	height: @onboardingDialogHeight;
	color: @color-base;
	// REVIEW Overwrite Codex's gutter between header/body and body/footer. The gutter
	// generated by the Cdx-dialog gap: 32px and position the paginator just below header
	// stylelint-disable-next-line selector-class-pattern
	&.cdx-dialog {
		gap: 0;
	}

	// Position the paginator relative to the dialog content to scroll with step content
	.ext-growthExperiments-OnboardingDialog__content {
		position: relative;

		&__paginator {
			top: 0;
		}
	}

	&__image {
		// REVIEW The negative margin values are needed to overwrite
		// CdxDialog body left and right gutter and display full bleed image
		margin-left: @spacing-150 * -1em;
		margin-right: @spacing-150 * -1em;
		background-color: @onboardingBackgroundColor;
		// This border color is not a DS border-color yet and may be changed
		// See https://phabricator.wikimedia.org/T332567
		border-bottom: @border-width-base @border-style-base @onboardingImageBorderColor;
		display: flex;
		justify-content: center;
	}

	&__title {
		padding-top: @spacing-50;
		font-size: @font-size-medium;
		font-weight: @font-weight-bold;
	}

	&__text {
		padding-top: @spacing-50;
		font-size: @font-size-small;
		line-height: @line-height-xx-small;

		&__label {
			color: @color-placeholder;
			font-size: @font-size-x-small;
			line-height: @line-height-xx-small;
			font-style: italic;
			margin-top: @spacing-50;
		}

		&__example {
			padding: @spacing-75;
			background: @background-color-interactive-subtle;
			// As discussed on https://phabricator.wikimedia.org/T332567
			// this color is not a DS border-color yet
			// and should be changed
			border: @border-width-base @border-style-base @onboardingExampleBorderColor;
			box-sizing: @box-sizing-base;
			border-radius: @border-radius-base;
			margin-bottom: @spacing-50;
			line-height: @line-height-xx-small;
			font-size: @font-size-small;

			&__highlighted {
				background-color: @background-color-progressive-subtle;
			}
		}

		&__link {
			.cdx-mixin-link-base();
		}

		&__list {
			ul {
				list-style: inside;
				margin-left: @spacing-50;

				li {
					margin-top: @spacing-50;
				}
			}
		}
	}
}

</style>
