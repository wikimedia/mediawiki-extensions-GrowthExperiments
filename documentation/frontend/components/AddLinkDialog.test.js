import { describe, expect, it } from 'vitest';
import { mount } from '@vue/test-utils';
import AddLinkDialog from './AddLinkDialog.vue';

describe( 'AddLinkDialog', () => {
	it( 'should render 3 step with Add a link on-boarding content', () => {
		const wrapper = mount( AddLinkDialog, {
			props: { open: true }
		} );
		expect( wrapper.html() ).toMatchSnapshot();
	} );
	it( 'should render second step with Add a link on-boarding content', () => {
		const wrapper = mount( AddLinkDialog, {
			props: {
				open: true,
				initialStep: 2
			}
		} );
		expect( wrapper.html() ).toMatchSnapshot();
	} );
	it( 'should render third step with Add a link on-boarding content', () => {
		const wrapper = mount( AddLinkDialog, {
			props: {
				open: true,
				initialStep: 3
			}
		} );
		expect( wrapper.html() ).toMatchSnapshot();
	} );
} );
