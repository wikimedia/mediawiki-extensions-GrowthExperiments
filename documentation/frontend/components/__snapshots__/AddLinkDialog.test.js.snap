// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddLinkDialog > should render 3 step with Add a link on-boarding content 1`] = `
"<!-- eslint-disable vue/no-v-model-argument -->
<!-- eslint-disable vue/no-v-model-argument -->
<transition-stub name=\\"cdx-dialog-fade\\" appear=\\"true\\" persisted=\\"false\\" css=\\"true\\">
  <div class=\\"cdx-dialog-backdrop\\">
    <div tabindex=\\"0\\"></div>
    <div class=\\"cdx-dialog cdx-dialog--horizontal-actions ext-growthExperiments-OnboardingDialog ext-growthExperiments-AddLinkDialog\\" role=\\"dialog\\" aria-label=\\"Introduction\\">
      <header class=\\"cdx-dialog__header\\">
        <div class=\\"ext-growthExperiments-OnboardingDialog__header\\">
          <!-- Slot for dialog title -->
          <h4 class=\\"ext-growthExperiments-OnboardingDialog__header__title\\"> Introduction </h4>
          <!-- Header button that will display in all steps except last one
				if slot with inner text is provided  -->
          <div class=\\"ext-growthExperiments-OnboardingDialog__header__button\\"><button class=\\"cdx-button cdx-button--action-default cdx-button--weight-quiet\\"> Skip all </button></div>
        </div>
      </header>
      <div class=\\"cdx-dialog-focus-trap\\" tabindex=\\"-1\\"></div>
      <div class=\\"cdx-dialog__body\\">
        <div class=\\"ext-growthExperiments-OnboardingDialog__content\\">
          <!-- Dialog paginator indicate current steps and total steps. -->
          <div class=\\"ext-growthExperiments-OnboardingPaginator ext-growthExperiments-OnboardingDialog__content__paginator\\">
            <div class=\\"ext-growthExperiments-OnboardingPaginator__dots\\">
              <div class=\\"ext-growthExperiments-OnboardingPaginator__dots__dot ext-growthExperiments-OnboardingPaginator__dots__dot--active\\"></div>
              <div class=\\"ext-growthExperiments-OnboardingPaginator__dots__dot\\"></div>
              <div class=\\"ext-growthExperiments-OnboardingPaginator__dots__dot\\"></div>
            </div>
            <p>1 of 3</p>
          </div>
          <div class=\\"ext-growthExperiments-OnboardingDialog__content__transition\\">
            <transition-stub name=\\"right\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
              <div>
                <!-- eslint-disable max-len -->
                <div role=\\"img\\" aria-label=\\"Illustration of the moon article, with the articles Earth and Satellite being suggested as links that could be added to the associated text.\\" class=\\"ext-growthExperiments-AddLinkDialog__image ext-growthExperiments-AddLinkDialog__image--1\\">
                  <!-- eslint-enable max-len -->
                </div>
                <h5 class=\\"ext-growthExperiments-AddLinkDialog__title\\"> Adding links will help people learn faster. </h5>
                <div class=\\"ext-growthExperiments-AddLinkDialog__text\\">
                  <p> You will decide whether words in one Wikipedia article should link to other Wikipedia articles. </p>
                  <div class=\\"ext-growthExperiments-AddLinkDialog__text__label\\"> Example sentence </div>
                  <div class=\\"ext-growthExperiments-AddLinkDialog__text__example\\"> The moon is the only
                    <!-- eslint-disable-next-line max-len --><mark class=\\"ext-growthExperiments-AddLinkDialog__text__example__highlighted\\"> natural satellite </mark> that <mark class=\\"ext-growthExperiments-AddLinkDialog__text__example__highlighted\\"> orbits </mark> around the <mark class=\\"ext-growthExperiments-AddLinkDialog__text__example__highlighted\\"> Earth </mark>.
                  </div>
                  <p> No special knowledge about the article is needed to do this task. </p>
                </div>
              </div>
            </transition-stub>
          </div>
        </div>
      </div>
      <footer class=\\"cdx-dialog__footer\\">
        <div class=\\"ext-growthExperiments-OnboardingDialog__footer\\">
          <div class=\\"ext-growthExperiments-OnboardingDialog__footer__actions\\">
            <!-- Footer Actions Prev -->
            <!-- The first step of the dialog displays a checkbox
					if text content for the checkbox slot  is provided. -->
            <div class=\\"ext-growthExperiments-OnboardingDialog__footer__actions-prev\\">
              <!-- eslint-disable-next-line max-len --><span class=\\"cdx-checkbox\\"><input id=\\"cdx-checkbox-1\\" class=\\"cdx-checkbox__input\\" type=\\"checkbox\\" value=\\"false\\"><span class=\\"cdx-checkbox__icon\\"></span><label class=\\"cdx-checkbox__label\\" for=\\"cdx-checkbox-1\\"> Don't show again </label></span>
            </div><!-- Footer Acctions Next -->
            <!-- The last step displays a button with a slot for the text content -->
            <!-- All the previous steps display an Icon only button to navigate
					to the next step -->
            <div class=\\"ext-growthExperiments-OnboardingDialog__footer__actions-next\\"><button class=\\"cdx-button cdx-button--action-progressive cdx-button--weight-primary cdx-button--framed cdx-button--icon-only cdx-button--icon-only\\" aria-label=\\"next\\"><span class=\\"cdx-icon cdx-icon--medium\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 20 20\\"><title>next</title><g><path d=\\"M7 1 5.6 2.5 13 10l-7.4 7.5L7 19l9-9z\\"></path></g></svg></span></button></div>
          </div>
        </div>
      </footer>
    </div>
    <div tabindex=\\"0\\"></div>
  </div>
</transition-stub>"
`;

exports[`AddLinkDialog > should render second step with Add a link on-boarding content 1`] = `
"<!-- eslint-disable vue/no-v-model-argument -->
<!-- eslint-disable vue/no-v-model-argument -->
<transition-stub name=\\"cdx-dialog-fade\\" appear=\\"true\\" persisted=\\"false\\" css=\\"true\\">
  <div class=\\"cdx-dialog-backdrop\\">
    <div tabindex=\\"0\\"></div>
    <div class=\\"cdx-dialog cdx-dialog--horizontal-actions ext-growthExperiments-OnboardingDialog ext-growthExperiments-AddLinkDialog\\" role=\\"dialog\\" aria-label=\\"Introduction\\">
      <header class=\\"cdx-dialog__header\\">
        <div class=\\"ext-growthExperiments-OnboardingDialog__header\\">
          <!-- Slot for dialog title -->
          <h4 class=\\"ext-growthExperiments-OnboardingDialog__header__title\\"> Introduction </h4>
          <!-- Header button that will display in all steps except last one
				if slot with inner text is provided  -->
          <div class=\\"ext-growthExperiments-OnboardingDialog__header__button\\"><button class=\\"cdx-button cdx-button--action-default cdx-button--weight-quiet\\"> Skip all </button></div>
        </div>
      </header>
      <div class=\\"cdx-dialog-focus-trap\\" tabindex=\\"-1\\"></div>
      <div class=\\"cdx-dialog__body\\">
        <div class=\\"ext-growthExperiments-OnboardingDialog__content\\">
          <!-- Dialog paginator indicate current steps and total steps. -->
          <div class=\\"ext-growthExperiments-OnboardingPaginator ext-growthExperiments-OnboardingDialog__content__paginator\\">
            <div class=\\"ext-growthExperiments-OnboardingPaginator__dots\\">
              <div class=\\"ext-growthExperiments-OnboardingPaginator__dots__dot ext-growthExperiments-OnboardingPaginator__dots__dot--active\\"></div>
              <div class=\\"ext-growthExperiments-OnboardingPaginator__dots__dot ext-growthExperiments-OnboardingPaginator__dots__dot--active\\"></div>
              <div class=\\"ext-growthExperiments-OnboardingPaginator__dots__dot\\"></div>
            </div>
            <p>2 of 3</p>
          </div>
          <div class=\\"ext-growthExperiments-OnboardingDialog__content__transition\\">
            <transition-stub name=\\"right\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
              <div>
                <!-- eslint-disable max-len -->
                <div role=\\"img\\" aria-label=\\"Illustration of the moon article, next to the article there's a robot representing a machine suggestion - asking if a word on the Moon article should be linked to the Cheese article.\\" class=\\"ext-growthExperiments-AddLinkDialog__image ext-growthExperiments-AddLinkDialog__image--2\\">
                  <!-- eslint-enable max-len -->
                </div>
                <h5 class=\\"ext-growthExperiments-AddLinkDialog__title\\"> Suggested links are machine-generated, and can be incorrect. </h5>
                <div class=\\"ext-growthExperiments-AddLinkDialog__text\\">
                  <p>
                    <!-- eslint-disable-next-line max-len --> The suggestions might be on words that donâ€™t need them, or might link to the wrong article. Use your judgment to decide whether they are right or wrong.
                  </p><a class=\\"ext-growthExperiments-AddLinkDialog__text__link\\" href=\\"\\"> Learn more about machine suggestions </a>
                </div>
              </div>
            </transition-stub>
          </div>
        </div>
      </div>
      <footer class=\\"cdx-dialog__footer\\">
        <div class=\\"ext-growthExperiments-OnboardingDialog__footer\\">
          <div class=\\"ext-growthExperiments-OnboardingDialog__footer__actions\\">
            <!-- Footer Actions Prev -->
            <!-- The first step of the dialog displays a checkbox
					if text content for the checkbox slot  is provided. -->
            <!-- All the following steps display an Icon only button to navigate
					to the previous step -->
            <div class=\\"ext-growthExperiments-OnboardingDialog__footer__actions-prev\\"><button class=\\"cdx-button cdx-button--action-default cdx-button--weight-normal cdx-button--framed cdx-button--icon-only\\" aria-label=\\"previous\\"><span class=\\"cdx-icon cdx-icon--medium\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 20 20\\"><title>previous</title><g><path d=\\"m4 10 9 9 1.4-1.5L7 10l7.4-7.5L13 1z\\"></path></g></svg></span></button></div><!-- Footer Acctions Next -->
            <!-- The last step displays a button with a slot for the text content -->
            <!-- All the previous steps display an Icon only button to navigate
					to the next step -->
            <div class=\\"ext-growthExperiments-OnboardingDialog__footer__actions-next\\"><button class=\\"cdx-button cdx-button--action-progressive cdx-button--weight-primary cdx-button--framed cdx-button--icon-only cdx-button--icon-only\\" aria-label=\\"next\\"><span class=\\"cdx-icon cdx-icon--medium\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 20 20\\"><title>next</title><g><path d=\\"M7 1 5.6 2.5 13 10l-7.4 7.5L7 19l9-9z\\"></path></g></svg></span></button></div>
          </div>
        </div>
      </footer>
    </div>
    <div tabindex=\\"0\\"></div>
  </div>
</transition-stub>"
`;

exports[`AddLinkDialog > should render third step with Add a link on-boarding content 1`] = `
"<!-- eslint-disable vue/no-v-model-argument -->
<!-- eslint-disable vue/no-v-model-argument -->
<transition-stub name=\\"cdx-dialog-fade\\" appear=\\"true\\" persisted=\\"false\\" css=\\"true\\">
  <div class=\\"cdx-dialog-backdrop\\">
    <div tabindex=\\"0\\"></div>
    <div class=\\"cdx-dialog cdx-dialog--horizontal-actions ext-growthExperiments-OnboardingDialog ext-growthExperiments-AddLinkDialog\\" role=\\"dialog\\" aria-label=\\"Introduction\\">
      <header class=\\"cdx-dialog__header\\">
        <div class=\\"ext-growthExperiments-OnboardingDialog__header\\">
          <!-- Slot for dialog title -->
          <h4 class=\\"ext-growthExperiments-OnboardingDialog__header__title\\"> Introduction </h4>
          <!-- Header button that will display in all steps except last one
				if slot with inner text is provided  -->
          <div class=\\"ext-growthExperiments-OnboardingDialog__header__button\\">
            <!--v-if-->
          </div>
        </div>
      </header>
      <div class=\\"cdx-dialog-focus-trap\\" tabindex=\\"-1\\"></div>
      <div class=\\"cdx-dialog__body\\">
        <div class=\\"ext-growthExperiments-OnboardingDialog__content\\">
          <!-- Dialog paginator indicate current steps and total steps. -->
          <div class=\\"ext-growthExperiments-OnboardingPaginator ext-growthExperiments-OnboardingDialog__content__paginator\\">
            <div class=\\"ext-growthExperiments-OnboardingPaginator__dots\\">
              <div class=\\"ext-growthExperiments-OnboardingPaginator__dots__dot ext-growthExperiments-OnboardingPaginator__dots__dot--active\\"></div>
              <div class=\\"ext-growthExperiments-OnboardingPaginator__dots__dot ext-growthExperiments-OnboardingPaginator__dots__dot--active\\"></div>
              <div class=\\"ext-growthExperiments-OnboardingPaginator__dots__dot ext-growthExperiments-OnboardingPaginator__dots__dot--active\\"></div>
            </div>
            <p>3 of 3</p>
          </div>
          <div class=\\"ext-growthExperiments-OnboardingDialog__content__transition\\">
            <transition-stub name=\\"right\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
              <div>
                <!-- eslint-disable max-len -->
                <div role=\\"img\\" aria-label=\\"Illustration of an article sentence showing link suggestions for two different words. Next to the suggestion there's a blue check icon and a red cross icon for the options to accept or reject the suggestion.\\" class=\\"ext-growthExperiments-AddLinkDialog__image ext-growthExperiments-AddLinkDialog__image--3\\">
                  <!-- eslint-enable max-len -->
                </div>
                <h5 class=\\"ext-growthExperiments-AddLinkDialog__title\\"> Guidelines </h5>
                <div class=\\"ext-growthExperiments-AddLinkDialog__text\\">
                  <div class=\\"ext-growthExperiments-AddLinkDialog__text__list\\">
                    <ul>
                      <li>Link concepts that a reader might want to learn more about.</li>
                      <li>Make sure the link is going to the right article.</li>
                      <li>Don't link common words, years, or dates.</li>
                      <li>If you're not sure, skip.</li>
                    </ul>
                  </div>
                </div>
              </div>
            </transition-stub>
          </div>
        </div>
      </div>
      <footer class=\\"cdx-dialog__footer\\">
        <div class=\\"ext-growthExperiments-OnboardingDialog__footer\\">
          <div class=\\"ext-growthExperiments-OnboardingDialog__footer__actions\\">
            <!-- Footer Actions Prev -->
            <!-- The first step of the dialog displays a checkbox
					if text content for the checkbox slot  is provided. -->
            <!-- All the following steps display an Icon only button to navigate
					to the previous step -->
            <div class=\\"ext-growthExperiments-OnboardingDialog__footer__actions-prev\\"><button class=\\"cdx-button cdx-button--action-default cdx-button--weight-normal cdx-button--framed cdx-button--icon-only\\" aria-label=\\"previous\\"><span class=\\"cdx-icon cdx-icon--medium\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"20\\" height=\\"20\\" viewBox=\\"0 0 20 20\\"><title>previous</title><g><path d=\\"m4 10 9 9 1.4-1.5L7 10l7.4-7.5L13 1z\\"></path></g></svg></span></button></div><!-- Footer Acctions Next -->
            <!-- The last step displays a button with a slot for the text content -->
            <div class=\\"ext-growthExperiments-OnboardingDialog__footer__actions-next\\"><button class=\\"cdx-button cdx-button--action-progressive cdx-button--weight-primary cdx-button--framed\\"> Get started </button></div>
          </div>
        </div>
      </footer>
    </div>
    <div tabindex=\\"0\\"></div>
  </div>
</transition-stub>"
`;
